### ISSUE TITLE
Add support for remote connections using SSH

### ISSUE THREAD

**Issue by albertvipe** [docker-java/docker-java#1130]
On the latest release of docker ([18.09](https://docs.docker.com/engine/release-notes/#1809)) they have added [support for remote connections using SSH](https://github.com/docker/cli/pull/1014).

It would be awesome to have it implemented and released on docker-java.

**Comment by jruhe-adesso** [comment]
ssh:// scheme still unsupported :(

**Comment by boxlady** [comment]
This would be nice... a more secure remote connection inside for inside PHPStorm.

**Comment by firexel** [comment]
Since docker-machine is slowly dying, this would be nice to have to upload builds

**Comment by pie-ai** [comment]
Mounting the remote socket looks good: https://medium.com/@dperny/forwarding-the-docker-socket-over-ssh-e6567cfab160

**Comment by mwiede** [comment]
I started a module based on jsch. Please have a look at it

https://github.com/mwiede/docker-java/blob/master/docker-java-transport-ssh/Readme.md

Some problems in direct attached stream and integration suite still exist.

**Comment by bsideup** [comment]
@mwiede nice! FYI in Testcontainers, we use the following dependency:
`org.glassfish.main.external:trilead-ssh2-repackaged:4.1.2`

It was used for port forwarding and worked pretty well :) you may want to try it in your prototype

**Comment by unikitty37** [comment]
I use RubyMine as an IDE. It uses this library to connect to Docker. I now need to access Docker containers on a remote machine, using SSH.

I can't do it from within the IDE, because this library doesn't support SSH. That means RubyMine throws all sorts of warnings about things not being installed, simply because it can't see the containers.

Is there *any* progress on this?

Failing that, is there a simple solution that works with RubyMine? The link posted by @pie-ai doesn't help, as entering `/path/to/docker.sock` as the engine API in RubyMine simply produces a NullPointerException. Entering `file:///path/to/docker.sock` just brings us back to "unsupported protocol scheme found" once more.

**Comment by trajano** [comment]
Does the SSH transport merely call the `docker` command on the remote or does it do something special to provide it access to the docker daemon?

**Comment by mwiede** [comment]
The SSH transport channel provides a socket to call the Docker API on the
host, not the Docker CLI. I described this setup on
https://github.com/mwiede/docker-java/blob/master/docker-java-transport-ssh/Readme.md

**Comment by unikitty37** [comment]
I'd hope that this issue wouldn't get auto-closed as stale!

I'd also hope that JetBrains might be able to take a look at fixing this, since they're making money off products using this library…

**Comment by bsideup** [comment]
@unikitty37 although JetBrains are indirectly making money with docker-java, there is currently no major investment into maintaining docker-java from their side, nor sponsor the current maintainers.

If anything to be done on this issue, you should expect it from the community (e.g. the efforts by @mwiede).
Unfortunately, due to my limited capacity, I wasn't able to look at the implementation yet, but I eventually will. Thanks for up'ing the issue 👍 I ACKed it, so that it won't get stale anymore.

**Comment by bsideup** [comment]
FYI https://github.com/docker-java/docker-java/pull/1818 removes the too restrictive protocol check in `DefaultDockerClientConfig`, so it becomes transport's concern whether some protocol is supported or not.

**Comment by mwiede** [comment]
Thanks @bsideup this should make it possible to publish ssh transport module of #1440 separate from this project.

**Comment by bsideup** [comment]
@mwiede yep, that was an intention - in the current form, it would be hard to accept it to docker-java's repo, but one can publish it separately, collect feedback, and, once it is stable enough, we could reconsider adding it to docker-java. Thank you!

**Comment by mwiede** [comment]
FYI I have released a module supporting ssh protocol at https://github.com/mwiede/docker-java-transport-jsch

**Comment by fsdrw08** [comment]
any update about this request?

**Comment by cvgaviao** [comment]
> FYI I have released a module supporting ssh protocol at https://github.com/mwiede/docker-java-transport-jsch

Is there a way to use it with TestContainers ?

**Comment by fostermi** [comment]
@mwiede how would one deploy/install/use the docker-java-transport-jsch module in a running Jenkins instance, if you wouldn't mind?
